File: src/main/java/arc/arc/invest/items/GenericItem.java
package arc.arc.invest.items;

import org.bukkit.inventory.ItemStack;

import java.util.Optional;

public abstract class GenericItem {

    public abstract ItemStack stack();

    public abstract boolean fits(ItemStack stack);

    public boolean isTool(){
        return false;
    }

    public static Optional<GenericItem> fromString(String s){
        NamedItem namedItem = NamedItem.find(s);
        System.out.println("Named item from "+s+" = "+namedItem);
        if(namedItem != null) return Optional.of(namedItem);
        SimpleItem simpleItem = SimpleItem.fromMaterial(s);
        if(simpleItem == null) return Optional.empty();

        return Optional.of(simpleItem);
    }

}

File: src/main/kotlin/ru/arc/invest/items/GenericItem.kt
package ru.arc.invest.items

import org.bukkit.inventory.ItemStack
import java.util.Optional

abstract class GenericItem {
    abstract fun stack(): ItemStack
    abstract fun fits(stack: ItemStack): Boolean
    open fun isTool(): Boolean = false

    companion object {
        @JvmStatic
        fun fromString(s: String): Optional<GenericItem> {
            val namedItem = NamedItem.find(s)
            println("Named item from $s = $namedItem")
            if (namedItem != null) return Optional.of(namedItem)
            val simpleItem = SimpleItem.fromMaterial(s) ?: return Optional.empty()
            return Optional.of(simpleItem)
        }
    }
}
